meilisearch-core/src/update/documents_addition.rs

```diff
@@ -253,22 +267,31 @@ pub fn reindex_all_documents(writer: &mut heed::RwTxn<MainT>, index: &store::Ind
     for document_id in documents_ids_to_reindex {
         for result in index.documents_fields.document_fields(writer, document_id)? {
             let (field_id, bytes) = result?;
-            let value: serde_json::Value = serde_json::from_slice(bytes)?;
+            let value: Value = serde_json::from_slice(bytes)?;
             ram_store.insert((document_id, field_id), value);
         }
 
-        for ((docid, field_id), value) in ram_store.drain() {
-            serialize_value_with_id(
-                writer,
-                field_id,
-                &schema,
-                docid,
-                index.documents_fields,
-                index.documents_fields_counts,
-                &mut indexer,
-                &mut ranked_map,
-                &value
-            )?;
+        // For each key-value pair in the document.
+        for ((document_id, field_id), value) in ram_store.drain() {
+            let serialized = serde_json::to_vec(&value)?;
+            index.documents_fields.put_document_field(writer, document_id, field_id, &serialized)?;
+
+            if let Some(indexed_pos) = schema.is_indexed(field_id) {
+                let number_of_words = index_value(&mut indexer, document_id, *indexed_pos, &value);
+                if let Some(number_of_words) = number_of_words {
+                    index.documents_fields_counts.put_document_field_count(
+                        writer,
+                        document_id,
+                        *indexed_pos,
+                        number_of_words as u16,
+                    )?;
+                }
+            }
+
+            if schema.is_ranked(field_id) {
+                let number = value_to_number(&value).unwrap_or_default();
+                ranked_map.insert(document_id, field_id, number);
+            }
```

Comment:
Proposal: Create an inline method for that. Its writer 2 times in the same file.